### PR TITLE
[feature](hudi) support hudi time travel in external table

### PR DESCRIPTION
**Pull request by AshinGau**
## Proposed changes

Support hudi time travel in external table:
```
select * from hudi_table for time as of '20230712221248';
```
PR(https://github.com/apache/doris/pull/15418) supports to take timestamp or version as the snapshot ID in iceberg, but hudi only has timestamp as the snapshot ID. Therefore, when querying hudi table with `for version as of`, error will be thrown like:
```
ERROR 1105 (HY000): errCode = 2, detailMessage = Hudi table only supports timestamp as snapshot ID
```
The supported formats of timestamp in hudi are: 'yyyy-MM-dd HH:mm:ss[.SSS]' or 'yyyy-MM-dd' or 'yyyyMMddHHmmss[SSS]', which is consistent with the [time-travel-query.](https://hudi.apache.org/docs/quick-start-guide#time-travel-query)

## Partitioning Strategies
Before this PR, hudi's partitions need to be synchronized to hive through [hive-sync-tool](https://hudi.apache.org/docs/syncing_metastore/#hive-sync-tool), or by setting very complex synchronization parameters in [spark conf](https://hudi.apache.org/docs/syncing_metastore/#sync-template). These processes are exceptionally complex and unnecessary, unless you want to query hudi data through hive.

In addition, partitions are changed in time travel. We cannot guarantee the correctness of time travel through partition synchronization.

So this PR directly obtain partitions by reading hudi meta information. Caching and updating table partition information through hudi instant timestamp, and reusing Doris' partition pruning.

### CODE REVIEW

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 53.16 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17167400626 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712103121_clickbench_pr_176790.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.29 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162119953 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712113757_clickbench_pr_176858.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.32 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17162797435 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712122202_clickbench_pr_176887.html

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 1]
```diff
@@ -128,7 +128,7 @@ private CatalogIf removeCatalog(long catalogId) {
             catalog.onClose();
             nameToCatalog.remove(catalog.getName());
             lastDBOfCatalog.remove(catalog.getName());
-            Env.getCurrentEnv().getExtMetaCacheMgr().removeCache(catalog.getName());
+            Env.getCurrentEnv().getExtMetaCacheMgr().removeCache(catalog.getId());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:131]: Is this a bug?

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiPartitionMgr.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    public void cleanPartitionProcess(long catalogId) {
+        HudiPartitionProcessor processor = partitionProcessors.get(catalogId);
+        if (processor != null) {
+            processor.cleanUp();
+        }
+    }
+
+    public void cleanDataBasePartitions(long catalogId, String dbName) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiPartitionMgr.java:62]:
```suggestion
    public void cleanDatabasePartitions(long catalogId, String dbName) {
```

**Review by morningman — COMMENTED** [review]
Update the document for new syntax

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java (hunk 1 above)
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:131]: Yes!!!

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiPartitionMgr.java (hunk 2 above)
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiPartitionMgr.java:62]: Done

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.02 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17162297192 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712233120_clickbench_pr_177490.html

**Comment by AshinGau** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by kaka11chen — APPROVED** [review]
LGTM

**Pull request merged**